x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizerTests.java

```diff
@@ -1226,11 +1230,268 @@ public boolean isIndexed(FieldAttribute.FieldName field) {
+            | eval similarity = v_cosine(dense_vector, [0, 255, 255])
+            | sort similarity desc, keyword asc
+            | limit 1
+            """;
+
+        LogicalPlan plan = localPlan(plan(query, allTypesAnalyzer), TEST_SEARCH_STATS);
+
+        // EsqlProject with all fields including similarity and keyword
+        var project = as(plan, EsqlProject.class);
+        assertTrue(Expressions.names(project.projections()).contains("similarity"));
+        assertTrue(Expressions.names(project.projections()).contains("keyword"));
+
+        var topN = as(project.child(), TopN.class);
+
+        var eval = as(topN.child(), Eval.class);
+        assertThat(eval.fields(), hasSize(1));
+        var alias = as(eval.fields().getFirst(), Alias.class);
+        assertThat(alias.name(), equalTo("similarity"));
+
+        // Check replaced field attribute
+        var fieldAttr = as(alias.child(), FieldAttribute.class);
+        assertThat(fieldAttr.fieldName().string(), equalTo("dense_vector"));
+        assertThat(fieldAttr.name(), startsWith("$$dense_vector$CosineSimilarity"));
+        var field = as(fieldAttr.field(), FunctionEsField.class);
+        var blockLoaderFunctionConfig = as(field.functionConfig(), DenseVectorFieldMapper.VectorSimilarityFunctionConfig.class);
+        assertThat(blockLoaderFunctionConfig.similarityFunction(), is(CosineSimilarity.SIMILARITY_FUNCTION));
+        assertThat(blockLoaderFunctionConfig.vector(), equalTo(new float[] { 0.0f, 255.0f, 255.0f }));
+
+        // MvExpand[keyword{f}#23,keyword{r}#32]
+        var mvExpand = as(eval.child(), MvExpand.class);
+        assertThat(Expressions.name(mvExpand.target()), equalTo("keyword"));
+
+        // Inner EsqlProject with the pushed down function
+        var innerProject = as(mvExpand.child(), EsqlProject.class);
+        assertTrue(Expressions.names(innerProject.projections()).contains("keyword"));
+        assertTrue(
+            innerProject.projections()
+                .stream()
+                .anyMatch(p -> (p instanceof FieldAttribute fa) && fa.name().startsWith("$$dense_vector$CosineSimilarity"))
         );
+
+        // EsRelation[test_all][$$dense_vector$CosineSimilarity$33{f}#33, !alias_in..]
+        var esRelation = as(innerProject.child(), EsRelation.class);
+        assertTrue(esRelation.output().contains(fieldAttr));
+    }
+
+    public void testVectorFunctionsWithDuplicateFunctions() {
+        assumeTrue("requires similarity functions", EsqlCapabilities.Cap.VECTOR_SIMILARITY_FUNCTIONS_PUSHDOWN.isEnabled());
+        String query = """
+            from test_all
+            | eval s1 = v_dot_product(dense_vector, [1.0, 2.0, 3.0]), s2 = v_dot_product(dense_vector, [1.0, 2.0, 3.0]) * 2 / 3
+            | eval s3 = v_dot_product(dense_vector, [1.0, 2.0, 3.0]) + 5, r1 = v_dot_product(dense_vector, [4.0, 5.0, 6.0])
+            | eval r2 = v_dot_product(dense_vector, [4.0, 5.0, 6.0]) + v_cosine(dense_vector, [4.0, 5.0, 6.0])
```

Comment:
I think you'll see more interesting results when we use the same dot product in different commands. E.g. one EVAL, one WHERE, on STATS.

We combine subsequent EVALs, so you don't see the effect of the same similarity function occuring in different commands.